src/lsm/manifest_level.zig

```diff
@@ -729,6 +734,336 @@ pub fn ManifestLevelType(
+            double_buffer: stdx.BoundedArray(*const TableInfo, max_coalesce_window_size),
+            /// A counter of total tables (invisible and visible) retrieved from the
+            /// inner iterator. After this hits `max_coalesce_tables_to_search` we will
+            /// no longer defer to the inner iterator.
+            tables_visited: u32,
+            /// This indicates that we've taken as many tables from the inner
+            /// iterator as we can (either because it is empty or we've called
+            /// it `tables_visited` times) and there are no more buffered tables
+            /// to return, so we should not return any new
+            /// `CoalesceWindowIterator`s in the `next` method.
+            end_of_iter: bool,
+
+            pub fn from_index(tables: *const Tables, start_index: u32) CoalesceTableIterator {
+                return CoalesceTableIterator{
+                    .inner = tables.iterator_from_index(start_index, .ascending),
+                    .buffer = .{},
+                    .double_buffer = .{},
+                    .tables_visited = 0,
+                    .end_of_iter = false,
+                };
+            }
+
+            pub fn next(it: *CoalesceTableIterator) ?CoalesceWindowIterator {
+                if (it.end_of_iter) {
+                    assert(it.buffer.empty() and it.double_buffer.empty());
+                    return null;
+                }
+
+                it.prepare_buffer();
+
+                return CoalesceWindowIterator{
+                    .parent = it,
+                    .window_size = 0,
+                };
+            }
+
+            /// Prepare `buffer` from the contents of `buffer` and `double_buffer`.
+            fn prepare_buffer(it: *CoalesceTableIterator) void {
+                assert(it.buffer.count() + it.double_buffer.count() <= max_coalesce_window_size);
+                if (!it.double_buffer.empty()) {
+                    // Anything still in `buffer` needs to be visited after `double_buffer`.
+                    // This can happen if the contents of `buffer` were not fully iterated
+                    // by a previous CoalesceWindowIterator.
+                    while (it.pop_buffer_queue()) |table| {
+                        it.push_double_buffer_queue(table);
+                    }
+                    // Promote double_buffer to buffer.
+                    std.mem.swap(
+                        stdx.BoundedArray(*const TableInfo, max_coalesce_window_size),
+                        &it.buffer,
+                        &it.double_buffer,
+                    );
+                }
+            }
+
+            // todo - this is a hot function and might profit from memcpy optimization
+            fn pop_buffer_queue(it: *CoalesceTableIterator) ?*const TableInfo {
+                if (it.buffer.empty()) {
+                    return null;
+                }
+
+                const retval = it.buffer.get(0);
```

Comment:
We use `result` to name return variables

Reply:
Done.